DOCS-2216: Add SparkFun Reference Station Setup Guide (#2974)
viamrobotics/docs: docs/components/movement-sensor/gps/set-up-base-station.md

---
title: "Set up a SparkFun RTK Reference Station"
linkTitle: "SparkFun RTK Reference Station Setup"
weight: 20
type: "docs"
description: "Follow this guide to set up your own SparkFun RTK base station, a high-precision GPS module designed for use with the RTK satellite navigation technique."

> images: ["/icons/components/imu.svg"]

Reviewer: ```suggestion
images: ["/icons/components/imu.svg"]
no_component: true
```

Reviewer: and then I think you need to put 

```
toc_hide: true
```

in the frontmatter of the movement-sensor models. 

and remove `hide_children: true` from the mvoement-sensor/_index.md file

I think that's how the camera one is done

Reviewer: oh and this should be moved out of the gps folder. GPS folder no longer exists...

Author: @npentrel Done, I think, it shows up the same as camera calibrate now.

RTK improves the precision of positional data from satellite-based positioning systems like the Global Positioning System (GPS).
When placed in a known location, the RTK Reference Station calculates and transmits correction data to RTK-enabled GPS receivers based on satellite signals it receives.

After following this guide, you can provide the mount point, username, password, and URL from your reference station's NTRIP server in your Viam machine config for the [`gps-nmea-rtk-pmtk`](/components/movement-sensor/gps-nmea-rtk-pmtk/) and [`gps-nmea-rtk-serial`](/components/movement-sensor/gps-nmea-rtk-serial/) models of movement sensor to allow a rover or other machine configured with your sensor to apply corrections from the reference station, enabling your machine to achieve sub-centimeter-level accuracy in positioning.

## Reserve an RTK2GO casting service

To distribute the correction information from your reference station, you should reserve an RTK2GO casting service for your station.

Follow the instructions on [New Base Station Reservation](http://rtk2go.com/sample-page/new-reservation/) to enter information about your reference station.

> Ensure that the email you provide here is the same as the email you use to authenticate to the [Viam app](app.viam.com).

Reviewer 2: @sguequierre oh sorry if it wasn't clear, I didn't mean the viam authentication email. I meant that when the user is configuring their GPS RTK, they will need to populate something called `ntrip_username` in https://docs.viam.com/components/movement-sensor/gps-nmea-rtk-serial/. That `ntrip_username` is going to be the email that they use in http://rtk2go.com/sample-page/new-reservation/. Hope that clears it up

Author: Ah ok thank you I was confused! Will update

Press the **MODE** button once to activate the menu.
Press the button again to step through the available modes.
Once highlighted, pause on **Base** to select this mode.

Because you configured the base station to be in **Survey-in** mode in the [update Base configuration step](#update-base-configuration), it will take your base station up to six minutes to go into **Xmitting RTCM** mode.
After a few minutes it goes into **Casting** mode.
Once you see **Casting** on the mode menu display screen, go to `rtk2go.com:2101` in your browser and find your mount point in the source table.

## Next steps: configure an RTK movement sensor

Open the configuration instructions for your RTK-enabled movement sensor model:

- [`gps-nmea-rtk-pmtk`](/components/movement-sensor/gps-nmea-rtk-pmtk/)
- [`gps-nmea-rtk-serial`](/components/movement-sensor/gps-nmea-rtk-serial/)

Follow the configuration instructions, providing the values you entered for your NTRIP server [Base configuration](#update-base-configuration) as the equal attributes in your Viam machine config:

- `ntrip_url`: your RTK2GO casting URL, comprised of **Caster Host** and **Caster Port** (like `rtk2go.com:2101`)

> - `ntrip_username`: your **Caster User**, which should be your email

Reviewer 2: Can we further specify that this email is the one that is used to setup the RTK reference station in http://rtk2go.com/sample-page/new-reservation/?

weight: 20
type: "docs"
description: "Follow this guide to set up your own SparkFun RTK base station, a high-precision GPS module designed for use with the RTK satellite navigation technique."
images: ["/icons/components/imu.svg"]
# SMEs: Susmita
---

Follow this guide to set up your [SparkFun RTK Reference Station](https://www.sparkfun.com/products/22429), a high-precision GPS module designed for use with the Real-Time Kinematic (RTK) satellite navigation technique.
RTK improves the precision of positional data from satellite-based positioning systems like the Global Positioning System (GPS).
When placed in a known location, the RTK Reference Station calculates and transmits correction data to RTK-enabled GPS receivers based on satellite signals it receives.

After following this guide, you can provide the URL from your reference station's fixed-position NTRIP server as the `ntrip_url` attribute in your Viam machine config for the `gps-nmea-rtk-pmtk` and `gps-nmea-rtk-serial` models of movement sensor to allow rovers configured with these sensors to receive both GPS signals and correction data and apply corrections to achieve centimeter-level accuracy in positioning.

> ## Reserve a RTK2GO casting service

Author: ah yes sorry

### Update System configuration

Look on the top of the RTK setup configure page and ensure that the firmware version of the base station is v4.0:

{{< imgproc src="/components/movement-sensor/sparkfun-rtk-firmware.png" alt="The top of the SparkFun RTK Setup configure page with the RTK Firmware version displayed as v4.0." resize="1000x" style="width:500px" >}}

This is important since as of April 2024, version 4.0 allows the base station to send [RTCM](https://www.xyht.com/gnsslocation-tech/rtcm/) corrections over WiFi.
To update the version of the firmware to 4.0, select the **System Configuration** dropdown and follow the steps provided in [Updating Firmware Over-The-Air](https://docs.sparkfun.com/SparkFun_RTK_Firmware/firmware_update/#updating-firmware-over-the-air).

### Update GNSS configuration

Select the **GNSS Configuration** dropdown on the RTK setup configure page.
This is what each option means and what value you should select:

**Measurement Rate:**

> - **In Hz:** The number of solutions or location ‘fixes’ per second. Anything above 4Hz may cause Bluetooth congestion and/or logging discrepancies.

Author: Sorry copy/paste. Going to remove these descriptions after talking to Susmita anyways, they are on webpage @npentrel

  - Note: The measurement rate is overridden to 1Hz when you configure the reference station to **Base** mode.
  - Default: 4Hz.
  - Limit: 0.000122 to 10Hz.

>     Here, select the default value of **4Hz**.

Author: I meant here as in on this dropdown, does that make sense? Will still change

>     Here, select the default value of **4Hz**.

Author: Going to remove after talking to Susmita @npentrel

**Survey-In:** If the precise location of a base station is not known it may be obtained by ‘surveying’ the location.
The base is fixed in one place and takes approximately 60 seconds worth of readings to obtain a best fit location based on the measurements. This method achieves ~30cm accurate position but can vary.
Increasing the minimum observation time or required mean deviation increases accuracy but only to a point.
Better accuracy is achieved with long-term logging and post processing.

- Default: 60s and 5.0m.
- Limits: 60 to 900s, 1.0 to 5.0m.
- Enable the **Survey-In** mode, and configure it as follows:
  - **Minimum observation time**: 60 (default)
  - **Required Mean 3D standard deviation (m)**: 5.00 (default)

**Fixed**: You only want the **Fixed** mode if your reference station will be stationary at a permanent location.

- Keep this deselected.

**Enable NTRIP Server:** You use this NTRIP server to upload the base's correction data to the casting service.
Other devices can then access the correction data using an NTRIP client.

- Default: Disabled.
- Enable this option. Enter the following credentials:
  - **Caster Host:** rtk2go.com
  - **Caster Port:** 2101

>   - **Caster User:** <test@test.org>

Author: @susmitaSanyal Should I instruct them to put in their own emails here? 

MD linter won't let me write test@test.org without the carrots due to empty url rule

Reviewer 2: yes, users can use their own email. I didn't know what to put so I used test lol

Author: Ok sweet thanks!

Your **Profile Configuration** should look like this:

{{< imgproc src="/components/movement-sensor/sparkfun-profile-configuration.png" alt="Example Profile Configuration." resize="900x" style="width:450px" >}}

### Save your configuration changes

Select **Save Configuration** on the RTK setup configure page and wait for your configurations to take effect.

## Start RTK2GO casting

Using the reference kit's **MODE** button, switch the mode back to the default **Base**.
Press the **MODE** button once to activate the menu.
Press the button again to step through the available modes.
Once highlighted, pause on **Base** to select this mode.

Because you configured the base station to be in **Survey-in** mode in the [update Base configuration step](#update-base-configuration), it will take your base station up to six minutes to go into **Xmitting RTCM** mode.
Then, after a few minutes it should go into **Casting** mode.
Once you see **Casting** on the mode menu display screen, go to `rtk2go.com:2101` in your browser and find your mount point in the source table.

## Next steps: configure an RTK movement sensor

> Provide this mount point as the `ntrip_url` attribute in your Viam machine config for the RTK-enabled models of movement sensor.

Author: @susmitaSanyal Is this accurate? will it be URL? a bit unsure-- if another step lmk!

Reviewer 2: ntrip_url is `rtk2go.com:2101` , mount point would look something like "MTPT" or "VIAM_BASE". This is something the user determines in step 1

Author: So how does mount point work? What do you need it for? Sorry for the confusion @susmitaSanyal

Reviewer 2: Sierra and I had a discussion in person about this.

> Click on the link to follow configuration instructions for each model:

Reviewer: I don't think they should click both links... ;)
```suggestion
Next, follow the configuration instructions for the desired model:
```

- [`gps-nmea-rtk-pmtk`](/components/movement-sensor/gps/gps-nmea-rtk-pmtk/)
- [`gps-nmea-rtk-serial`](/components/movement-sensor/gps/gps-nmea-rtk-serial/)

> This allows rovers configured with these sensors to receive both GPS signals and correction data and apply corrections to achieve centimeter-level accuracy in positioning.

Reviewer: I think this should go with the sentence about providing the casting URL, there it's very disconnected and leaving me asking what "This" is
